tests/frontend/test_parameters.py

```diff
@@ -206,3 +229,124 @@ def test_parameter_unary_negation(self, p):
+        phi = prog.params('phi')
+
+        with prog as q:
+            gate(r, phi) | q[0]
+
+        # create numeric values
+        mapping = {'r': tf.Variable(0.1), 'phi': tf.Variable(0.1)}
+        prog.bind_params(mapping)
+
+        # verify bound parameters are correct
+        assert prog.free_params['r'].val is mapping['r']
+        assert prog.free_params['phi'].val is mapping['phi']
+
+        # assert executed program is constructed correctly
+        global applied
+        cmds = []
+        applied = []
+
+        with monkeypatch.context() as m:
+            m.setattr(sf.LocalEngine, "_run_program", mock_run_prog)
+            eng, _ = setup_eng(1)
+            result = eng.run(prog, args=mapping)
+
+        assert isinstance(applied[0].op, gate)
+        assert applied[0].op.p[0].val == mapping["r"]
+        assert applied[0].op.p[1].val == mapping["phi"]
+
+    @pytest.mark.parametrize("gate", [sf.ops.BSgate, sf.ops.S2gate])
+    def test_two_mode_gate_complex_phase(self, setup_eng, gate, monkeypatch):
+        """Test two mode gates with complex phase arguments"""
+        prog = sf.Program(2)
+
+        # create symbolic parameters
+        r = prog.params('r')
+        phi = prog.params('phi')
+
+        with prog as q:
+            gate(r, phi) | q
+
+        # create numeric values
+        mapping = {'r': tf.Variable(0.1), 'phi': tf.Variable(0.1)}
+        prog.bind_params(mapping)
+
+        # verify bound parameters are correct
+        assert prog.free_params['r'].val is mapping['r']
+        assert prog.free_params['phi'].val is mapping['phi']
+
+        # assert executed program is constructed correctly
+        global applied
+        cmds = []
+        applied = []
+
+        with monkeypatch.context() as m:
+            m.setattr(sf.LocalEngine, "_run_program", mock_run_prog)
+            eng, _ = setup_eng(2)
+
+            # The tensorflow backend doesn't natively support two-mode squeezing.
+            # Here, we mock the tf compiler and the backend to allow this program to run.
+            m.setattr(sf.Program, "compile", lambda self, *args, **kwargs: self)
+            eng.backend.two_mode_squeeze = lambda self, *args: self
+
+            result = eng.run(prog, args=mapping)
+
+        assert isinstance(applied[0].op, gate)
+        assert applied[0].op.p[0].val == mapping["r"]
+        assert applied[0].op.p[1].val == mapping["phi"]
+
+    def test_zgate(self, setup_eng, hbar, monkeypatch):
+        """Test the momentum displacement gate"""
+        prog = sf.Program(1)
+
+        # create symbolic parameters
+        p = prog.params('p')
+
+        with prog as q:
+            sf.ops.Zgate(p) | q[0]
+
+        # create numeric values
+        mapping = {'p': tf.Variable(0.1)}
+        prog.bind_params(mapping)
+
+        # verify bound parameters are correct
+        assert prog.free_params['p'].val is mapping['p']
+
+        # assert executed program is constructed correctly
+        global applied
```

Comment:
how come a `global` is used here?

Reply:
Was wondering the same thing myself! It essentially seems to help with having a module-level object, so the `applied` objects within the tests that are defined as `global applied` will be referencing the same object as the one that is used in this mock function. It is a way of switching from a local scope, though I'd imagine that this might have caveats in between tests if not torn down correctly.

Reply:
This is a bit of a hack from my end; while coding the tests, I realised that a global variable would allow me to 'extract' the internal engine command list with some monkeypatching 😆 I'll rewrite this as a fixture, which will hopefully avoid issues with set up and tear down.

Reply:
Have replaced this with a fixture :)